tests/unittest_regrtest.py

```diff
@@ -399,5 +401,33 @@ class Another(subclass):
     parse(code)
 
 
+def test_max_inferred_for_complicated_class_hierarchy() -> None:
+    """Regression test for a crash reported in https://github.com/PyCQA/pylint/issues/5679.
+
+    The class hierarchy of 'sqlalchemy' is so intricate that it becomes uninferable with
+    the standard max_inferred of 100. We used to crash when this happened.
+    """
+    # Create module and get relevant nodes
+    module = resources.build_file(
+        str(resources.RESOURCE_PATH / "max_inferable_limit_for_classes" / "main.py")
+    )
+    init_attr_node = module.body[-1].body[0].body[0].value.func
+    init_object_node = module.body[-1].mro()[-1]["__init__"]
+    super_node = next(init_attr_node.expr.infer())
+
+    # Arbitrarily limit the max number of infered nodes per context
+    InferenceContext.max_inferred = -1
+    context = InferenceContext()
+
+    # Try to infer 'object.__init__' > because of limit is impossible
+    for _ in bases._infer_stmts([init_object_node], context, frame=super):
+        pass
+
```

Comment:
Shouldn't there also be an assert here ?

Reply:
We could, it's just here to populate the `context` like it would normally. You can make a suggestion if you want to add it. I think `assert _ == Uninferable` should pass.

Reply:
I thought it was a test in itself because the  ``max_inferred`` value is changed before ? Why is there a particular value for max_inferred in the context if we only want to populate context ?

Reply:
When `max_inferred` is lower than the current number of `inferred` values in a context when return `Uninferable`: if we need to infer 101 nodes to understand what node A is then we let node A be `Uninferable`. This limit is reached in `pylint` for this particular inheritance pattern, but I couldn't get it to hit for `astroid` when only inferring the base files. (When running `pylint` other values get stored in the cache as well which is why we hit it sooner).
To make this bug consistent I set the limit to -1 so that we will always hit it, thereby replicating the behaviour in `pylint`.

Because we hit the limit we store yield `Uninferable` for node A and store this in the cache. However, without this commit we wouldn't store `Uninferable` for node A, but rather `[]`. Thus when we did another lookup for node A we would yield from `[]` instead of from `[Uninferable,]`. This made it so that while we had node A as a key in the inference cache we didn't actually have any values for that key, which caused the exception to be raised.

Reply:
Ok, this is what I understood as well. So would it make sense that the max_inferred limit worked as we thought it would and that we do yield an ``Uninferable`` ?

Reply:
Yes! `for _ in bases._infer_stmts([init_object_node], context, frame=super):` will work irrespective of this change. It is the `getattr` call that is/was causing problems. Because `getattr` also calls `_infer_stmts` it tries to use the cache and then yields from `[]`.

We could add the `assert _ == Uninferable` test, but it is not necessarily related to the actual bug here, which is why I originally excluded it. `_infer_stmts` on a "101-inference node" called once works fine, `_infer_stmts` called twice on a "101-inference node" is broken.